tests/spec/corpus.rs

```diff
@@ -0,0 +1,226 @@
+                "{}",
+                description
+            );
+        }
+
+        // native_to_canonical_extended_json( json_to_native(cEJ) ) = cEJ
+
+        let json_to_native_cej: Bson = cej.clone().into();
+
+        let native_to_canonical_extended_json_bson_to_native_cej =
+            json_to_native_cej.clone().into_canonical_extjson();
+
+        if !description.contains("1.2345678921232E+18") {
+            assert_eq!(
+                native_to_canonical_extended_json_bson_to_native_cej, cej,
+                "{}",
+                description,
+            );
+        }
+
+        // native_to_bson( json_to_native(cEJ) ) = cB (unless lossy)
+
+        if valid.lossy != Some(true) {
+            let mut native_to_bson_json_to_native_cej = Vec::new();
+            bson::encode_document(
+                &mut native_to_bson_json_to_native_cej,
+                json_to_native_cej.as_document().unwrap(),
+            )
+            .unwrap();
+
+            if test.bson_type != "0x13" {
+                assert_eq!(
+                    hex::encode(native_to_bson_json_to_native_cej).to_lowercase(),
+                    valid.canonical_bson.to_lowercase(),
+                    "{}",
+                    description,
+                );
+            }
+        }
+
+        // native_to_bson( bson_to_native(dB) ) = cB
+
+        if let Some(db) = valid.degenerate_bson {
+            let bson_to_native_db =
+                bson::decode_document(&mut hex::decode(&db).expect(&description).as_slice())
+                    .expect(&description);
+
+            let mut native_to_bson_bson_to_native_db = Vec::new();
+            bson::encode_document(&mut native_to_bson_bson_to_native_db, &bson_to_native_db)
+                .unwrap();
+
+            assert_eq!(
+                hex::encode(native_to_bson_bson_to_native_db).to_lowercase(),
+                valid.canonical_bson.to_lowercase(),
+                "{}",
+                description,
+            );
+        }
+
+        if let Some(ref degenerate_extjson) = valid.degenerate_extjson {
+            let dej: serde_json::Value =
+                serde_json::from_str(degenerate_extjson).expect(&description);
+
+            let json_to_native_dej: Bson = dej.clone().into();
+
+            // native_to_canonical_extended_json( json_to_native(dEJ) ) = cEJ
+
+            let native_to_canonical_extended_json_json_to_native_dej =
+                json_to_native_dej.clone().into_canonical_extjson();
+
+            if test.bson_type != "0x13" {
+                assert_eq!(
+                    native_to_canonical_extended_json_json_to_native_dej, cej,
+                    "{}",
+                    description,
+                );
+            }
+
+            // native_to_bson( json_to_native(dEJ) ) = cB (unless lossy)
+
+            if valid.lossy != Some(true) {
+                let mut native_to_bson_json_to_native_dej = Vec::new();
+                bson::encode_document(
```

Comment:
this seems like a helpful method to have on `Document` iteself. Like `to_vec` or `to_bytes` or something.

Reply:
I'm ambivalent about putting it in a public API; documents can get pretty large, and giving an API that's generic over `Write` (like `encode_document is) is generally the more idiomatic way to do something like this. Either way, I don't think it's a blocker for 1.0.

Reply:
I was more referring to the namespacing than the output type. It seems strange to me that `encode_document` and `decode_document` are both free floating functions in the `bson` module rather than methods under the `Document` namespace it self.

e.g.
```rust
json_to_native_dej.as_document().encode(&mut native_to_bson_json_to_native_dej).unwrap()
let doc = Document::decode(&mut native_to_bson_json_to_native_dej).unwrap()
```
The current free floating way seems very C like to me, whereas this feels more in line with Rust.

Reply:
I think it's intended to provide a similar API to `serde_json`, which uses free-floating functions for these things: https://docs.rs/serde_json/1.0.53/serde_json/#functions

Reply:
Those make sense as free floating because they're generic on the type being encoded/decoded; they can't be methods by nature of that. In this case, we have a concrete `Document` which we're encoding/decoding.

This convinces me further that, if we are to retain this method's functionality in the public API, we should restrict it to the `Document` namespace. That way, if/when we ever decide to match `serde_json`'s functionaltiy wrt to these pure generic functions (e.g. by adding `from_reader`), we won't have to deprecate the now redundant `from_document`.

This suggestion at this point certainly is out of scope for this PR, so I filed [RUST-429](https://jira.mongodb.org/browse/RUST-429) to discuss it further.

Reply:
Discussed offline